File: streamlit_app/screens/lobby.py
import streamlit as st

from game import round_flow


def render(ctx):
    room = ctx["room"]
    players = ctx["players"]
    uid = ctx["uid"]
    needed = room["config"]["playerCount"]
    is_host = room["hostUid"] == uid

    st.subheader("Waiting Room")

    base_url = st.secrets.get("app_base_url", "")
    if base_url:
        invite_link = f"{base_url.rstrip('/')}/?room={ctx['room_code']}"
        st.text_input("Invite link", value=invite_link, disabled=True, key="invite_link_display")
    else:
        st.caption("Share the room code above with the other players.")

    penalty_note = (
        "The winner can write a custom penalty."
        if room["config"].get("customPenalties")
        else "Playing for points only."
    )
    st.write(f"Needs exactly **{needed}** players. {penalty_note}")

    for p in players:
        dot = "🟢" if p.get("connected") else "⚪"
        host_tag = " :blue[host]" if p.get("isHost") else ""
        you_tag = " :orange[(you)]" if p["uid"] == uid else ""
        st.write(f"{dot} {p['displayName']}{you_tag}{host_tag}")

    if is_host:
        can_start = len(players) == needed
        label = "Start Game" if can_start else f"Waiting for {needed - len(players)} more player(s)…"
        if st.button(label, type="primary", disabled=not can_start, key="start_round_btn"):
            ctx["call"](round_flow.start_round, ctx["room_code"])
    else:
        st.caption("Waiting for the host to start the game…")

File: public/screens/lobby.js
export function renderLobby(root, ctx) {
  const { room, players, uid } = ctx;
  const isHost = room.hostUid === uid;
  const needed = room.config.playerCount;
  const canStart = players.length === needed;

  const inviteLink = ctx.inviteLinkFor(ctx.roomCode);

  root.innerHTML = `
    <div class="card-panel">
      <h2>Waiting Room</h2>
      <p class="muted">
        Share the room code above, or send an invite link. Needs exactly <strong>${needed}</strong> players.
        The pressure level (how many of the slowest slappers get penalized) is decided by which cards complete the win.
        ${room.config.customPenalties ? "The winner also gets to write a custom penalty." : "Playing for points only."}
      </p>
      <label for="invite-link">Invite link</label>
      <div class="invite-row">
        <input id="invite-link" readonly value="${escapeHtml(inviteLink)}" />
        <button id="copy-link" class="secondary">Copy</button>
      </div>
      <ul class="roster">
        ${players
          .map(
            (p) => `
          <li>
            <span>
              <span class="dot ${p.connected ? "online" : "offline"}"></span>
              ${escapeHtml(p.displayName)}
              ${p.uid === uid ? '<span class="you-tag">(you)</span>' : ""}
              ${p.isHost ? '<span class="host-tag">host</span>' : ""}
            </span>
          </li>`
          )
          .join("")}
      </ul>
      <div class="actions">
        ${
          isHost
            ? `<button id="start-round" ${canStart ? "" : "disabled"}>
                 ${canStart ? "Start Game" : `Waiting for ${needed - players.length} more player(s)…`}
               </button>`
            : `<p class="muted">Waiting for the host to start the game…</p>`
        }
      </div>
    </div>
  `;

  const startBtn = root.querySelector("#start-round");
  if (startBtn) {
    startBtn.addEventListener("click", async () => {
      startBtn.disabled = true;
      try {
        await ctx.call("start_round", { room_code: ctx.roomCode });
      } catch (_) {
        startBtn.disabled = !canStart;
      }
    });
  }

  root.querySelector("#copy-link").addEventListener("click", async () => {
    try {
      await navigator.clipboard.writeText(inviteLink);
      ctx.showToast("Invite link copied!");
    } catch (_) {
      root.querySelector("#invite-link").select();
      ctx.showToast("Couldn't auto-copy - link is selected, press Ctrl/Cmd+C.");
    }
  });
}

function escapeHtml(value) {
  return String(value).replace(/[&<>"']/g, (c) => ({
    "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;",
  }[c]));
}
